### DISCUSSION TITLE
Why cannot we rename the "master" branch to "main"?

### DISCUSSION THREAD

**Discussion by ehyoitsdavid** [curl/curl#10131] [Q&A]
Many other FOSS repos do that, but why does `curl/curl` repo not do that?

By renaming the name is more neutral

**Comment by bagder** [comment]
You are the first person to suggest we do. It's a pretty disruptive move which might be a reason why no one has proposed this before.

**Comment by ehyoitsdavid** [comment]
I'm sorry if I offended you.

I understand it is not a easy move for you guys, especially for automatic build scripts and CIs.

But as far as I can see some repos on GitHub have transitioned to `main` branch instead of `master` branch (putting politic correctness aside)

I am rather surprised that no one has bring this up until now (GitHub made that change in 2020)

It is two commands only: 
`git branch --move master main`
`git push --set-upstream origin main`

According to Pro Git:
> Changing the name of a branch like master/main/mainline/default will break the
integrations, services, helper utilities and build/release scripts that your repository
uses. Before you do this, **make sure you consult with your collaborators.** Also,
make sure you do a thorough search through your repo and update any references
to the old branch name in your code and scripts.

**Reply by andy5995** [reply]
> I'm sorry if I offended you.
> 
> I understand it is not a easy move for you guys, especially for automatic build scripts and CIs.
> 
> But as far as I can see some repos on GitHub have transitioned to `main` branch instead of `master` branch (putting politic correctness aside)

Most new repos created now are created with the "main" branch instead of master. But most repos that already existed with a "master" branch have not changed it to main.

> 
> I am rather surprised that no one has bring this up until now (GitHub made that change in 2020)
> 
> It is two commands only: `git branch --move master main` `git push --set-upstream origin main`

And everyone who has the branch cloned would need to update their local clones to match. And docs mentioning it would need to be changed as well. In my opinion, not worth it. All the new repos I create now use "trunk" as a default branch. I haven't changed the master branch of any of my pre-existing repositories.

> 
> According to Pro Git:
> 
> > Changing the name of a branch like master/main/mainline/default will break the
> > integrations, services, helper utilities and build/release scripts that your repository
> > uses. Before you do this, **make sure you consult with your collaborators.** Also,
> > make sure you do a thorough search through your repo and update any references
> > to the old branch name in your code and scripts.

As you can see, while it's easy to change/move the branch, many other things would be affected.

**Comment by dfandrich** [comment]
"Almost all" is certainly hyperbole. In my experience browsing GitHub in the last year or two, I think I've noticed one repo with main instead of master. Changing it is a big annoyance with zero benefits.

**Reply by xquery** [reply]
Well... both github and gitlab now default any new repos to **main** for obvious reason - saying it is a big annoyance with zero benefits is unhelpful as it misses the point (which is not technical). You might have meant that there are no technical benefits ... if so I disagree ... making these kind changes might find some bugs/assumptions that is baked in that we could make dynamic to improve the project.

overall +1  with cycles to contribute

**Reply by bagder** [reply]
We can argue about benefits, but we certainly know that there is a non-trivial amount of work and friction such a move will require and cause among a large number of developers. The curl org has **13** repositories, and I believe they all use 'master'.

**Reply by andy5995** [reply]
> Well... both github and gitlab now default any new repos to **main** for obvious reason - saying it is a big annoyance with zero benefits is unhelpful as it misses the point (which is not technical). You might have meant that there are no technical benefits ... if so I disagree ... making these kind changes might find some bugs/assumptions that is baked in that we could make dynamic to improve the project.
> 
> overall +1 with cycles to contribute

> Well... both github and gitlab now default any new repos to **main** for obvious reason - saying it is a big annoyance with zero benefits is unhelpful as it misses the point (which is not technical).

First I'll point out that I'm a minor curl contributor, not one of the maintainers.

I don't see that the reasons are obvious, but maybe I'm missing some information. I first read about the GitHub announcement to change the default branch to 'main' in 2020. My impression was that it was a public relations move with very little benefit (for example, I don't recall any corporate executives speaking out against political leaders who incited hatred and violence toward black people). In the context of a git branch, "master" is not at all related to the enslavement of people. And I hadn't heard that any black people had actually complained about the use of the term "master" in relation to git. But I don't hear and read everything, so please correct me with some informative links if my opinions don't seem to be based on enough facts. If I had some evidence that changing the names of all my master branches would prevent just a little of the evil, hatred, and bigotry in the world, I'd do it today, but I've never seen anything in the news that supported the idea that git master branches were contributing to racism.

**Reply by xquery** [reply]
I understand your pov I just do not agree ... I do agree that 'theatre' of any kind can be difficult to participate in as it is hard to see any kind of change ... changing repo names is never going to change the world, but we often need to start small to do that ...  and this is one of the very small (and easy) ways to communicate clearly to broader community that at a minimum we care enough to to think about change.

I have a unique view eg. I spent a lot of my younger life being the minority (being white, male) and seeing on a daily basis the corrosive, constant, ever present battle minorities (and women) have to endure just as 'baseline'... it is hard to grok  ... without at all intending to be parochial or implying anything on your informed view I would recommend reading https://en.wikipedia.org/wiki/Black_Like_Me

As much as I would like to never discuss such things in a issue tracker ... thank you for indulging me and thanks for listening/responding

**Reply by andy5995** [reply]
> I understand your pov I just do not agree ... I do agree that 'theatre' of any kind can be difficult to participate in as it is hard to see any kind of change ... changing repo names is never going to change the world, but we often need to start small to do that ... 

I agree that small actions can make a difference to huge problems.

>and this is one of the very small (and easy) ways to communicate clearly to broader community that at a minimum we care enough to to think about change.

I think treating people who use GitHub, and participate in GitHub projects, with respect regardless of their race, religion, background, or sexual orientation is an excellent way of communicating that we care enough to think about change. I think treating people as fellow humans far outweighs the benefits of branch renaming.

I see that essentially we have very similar views and I applaud your efforts to be an advocate.

> 
> I have a unique view eg. I spent a lot of my younger life being the minority (being white, male) and seeing on a daily basis the corrosive, constant, ever present battle minorities (and women) have to endure just as 'baseline'... it is hard to grok ... without at all intending to be parochial or implying anything on your informed view I would recommend reading https://en.wikipedia.org/wiki/Black_Like_Me

Thank you for the book recommendation. Courageous people such as Griffin are inspiring! (One of my favorites is "10 Days in a Madhouse" by Nellie Bly)

> As much as I would like to never discuss such things in a issue tracker

Maybe that's not a bad thing. The discussion is apparently still relevant, and if it didn't happen now, it would have happened in the future.

**Reply by cmeister2** [reply]
> We can argue about benefits, but we certainly know that there is a non-trivial amount of work and friction such a move will require and cause among a large number of developers. The curl org has **13** repositories, and I believe they all use 'master'.

I'm all in favour for switching repositories over to `main` where they aren't going to cause major disruption. Honestly I was considering it for the fuzzer repository recently.

I'd love to quantify the potential disruption though for the main repo. Maybe it's something to consider when 8.0 comes in, unless you're hoping for 8.0 to just be another release.
